IdentityCore/src/validation/MSIDCIAMAuthority.m

```diff
@@ -0,0 +1,187 @@
+// THE SOFTWARE.  
+
+#if !EXCLUDE_FROM_MSALCPP
+
+#import "MSIDCIAMAuthority.h"
+#import "MSIDCIAMAuthorityResolver.h"
+#import "MSIDTelemetryEventStrings.h"
+#import "MSIDAuthority+Internal.h"
+#import "MSIDJsonSerializableFactory.h"
+#import "MSIDJsonSerializableTypes.h"
+#import "MSIDProviderType.h"
+#import "MSIDAADTenant.h"
+#import "MSIDAADAuthority.h"
+#import "NSString+MSIDExtensions.h"
+
+@implementation MSIDCIAMAuthority
+
++ (void)load
+{
+    [MSIDJsonSerializableFactory registerClass:self forClassType:MSID_JSON_TYPE_CIAM_AUTHORITY];
+    [MSIDJsonSerializableFactory mapJSONKey:MSID_PROVIDER_TYPE_JSON_KEY keyValue:MSID_JSON_TYPE_PROVIDER_CIAM kindOfClass:MSIDAuthority.class toClassType:MSID_JSON_TYPE_CIAM_AUTHORITY];
+}
+
+- (nullable instancetype)initWithURL:(nonnull NSURL *)url
+                      validateFormat:(BOOL)validateFormat
+                           rawTenant:(nullable NSString *)rawTenant
+                             context:(nullable id<MSIDRequestContext>)context
+                               error:(NSError **)error
+{    
+    self = [self initWithURL:url validateFormat:validateFormat context:context error:error];
+    if (self)
+    {
+        if (rawTenant)
+        {
+            if ([self.class isAuthorityFormatValid:url context:context error:nil])
+            {
+                for(int a = 0; a < url.pathComponents.count; a = a + 1 ) {
+                    NSLog(@"value of path component %d: %@\n", a, url.pathComponents[a]);
+                   }
+                _url = [NSURL URLWithString:[NSString stringWithFormat:@"https://%@/%@", [url msidHostWithPortIfNecessary], rawTenant]];
+                _realm = rawTenant;
+            }
+        }
+    }
+
+    return self;
+}
+
+- (nullable instancetype)initWithURL:(NSURL *)url
+                      validateFormat:(BOOL)validateFormat
+                             context:(id<MSIDRequestContext>)context
+                               error:(NSError **)error
+{
+    self = [super initWithURL:url validateFormat:validateFormat context:context error:error];
+    
+    NSArray *hostComponents = [url.msidHostWithPortIfNecessary componentsSeparatedByString:@"."];
+
+    //If we have the URL tenant.ciamlogin.com
+    if (url.pathComponents.count == 0)
+    {
+        url = [url URLByAppendingPathComponent:hostComponents[0]];
+        url = [NSURL URLWithString:[url.absoluteString stringByAppendingString:@".onmicrosoftonline.com"]];
+    }
+    
+    if (self)
+    {
+        _url = url;
+        _url = [MSIDAADAuthority normalizedAuthorityUrl:url context:context error:error];
+        if (!_url) return nil;
+        self.url = url;
+    }
+    
+    return self;
+}
+    
+- (instancetype)initWithURL:(NSURL *)url
+                    context:(id<MSIDRequestContext>)context
+                      error:(NSError **)error
+{
+    return [self initWithURL:url validateFormat:YES context:context error:error];
+}
+
++ (BOOL)isAuthorityFormatValid:(NSURL *)url
```

Comment:
Can you clarify how calling handle error/ result here? Should calling terminated the flow early if it see error or move forward try next Authority class?

Reply:
Sure thing, the error handling here makes sure the provided URL/authority is in the correct format for an MSIDAuthority (making sure the realm can be extracted correctly, url scheme is "https"). It then checks to make sure that the guidelines are followed for creating a valid MSIDCIAMAuthority (URL has at least 3 segments in the path (i.e. https://<tenant>.ciamlogin.com...), host="ciamlogin"). If any of these authority validation conditions are not met, then the flow is terminated by an error.

Reply:
Thanks for explaining.